From 9f2eefb35a66097659b7ac632ee8f1d185d9b728 Mon Sep 17 00:00:00 2001
From: Sachin Prabhu <sprabhu@redhat.com>
Date: Mon, 28 Jul 2014 12:17:02 -0400
Subject: [fs] cifs: Fix memory leaks in SMB2_open

Message-id: <1406549822-10376-70-git-send-email-sprabhu@redhat.com>
Patchwork-id: 86766
O-Subject: [RHEL7.1 PATCH 69/69] [bz1082049]CIFS: Fix memory leaks in SMB2_open
Bugzilla: 1082049
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Pavel Shilovsky <pshilovsky@samba.org>

Cc: <stable@vger.kernel.org> # v3.12+
Signed-off-by: Pavel Shilovsky <pshilovsky@samba.org>
Reviewed-by: Shirish Pargaonkar <spargaonkar@suse.com>
Signed-off-by: Steve French <smfrench@gmail.com>
(cherry picked from commit 663a962151593c69374776e8651238d0da072459)
Signed-off-by: Sachin Prabhu <sprabhu@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/cifs/smb2pdu.c b/fs/cifs/smb2pdu.c
index b0037b6..b0b260d 100644
--- a/fs/cifs/smb2pdu.c
+++ b/fs/cifs/smb2pdu.c
@@ -1097,6 +1097,7 @@ SMB2_open(const unsigned int xid, struct cifs_open_parms *oparms, __le16 *path,
  int rc = 0;
  unsigned int num_iovecs = 2;
  __u32 file_attributes = 0;
+ char *dhc_buf = NULL, *lc_buf = NULL;
 
  cifs_dbg(FYI, "create/open\n");
 
@@ -1163,6 +1164,7 @@ SMB2_open(const unsigned int xid, struct cifs_open_parms *oparms, __le16 *path,
    kfree(copy_path);
    return rc;
   }
+  lc_buf = iov[num_iovecs-1].iov_base;
  }
 
  if (*oplock == SMB2_OPLOCK_LEVEL_BATCH) {
@@ -1177,9 +1179,10 @@ SMB2_open(const unsigned int xid, struct cifs_open_parms *oparms, __le16 *path,
   if (rc) {
    cifs_small_buf_release(req);
    kfree(copy_path);
-   kfree(iov[num_iovecs-1].iov_base);
+   kfree(lc_buf);
    return rc;
   }
+  dhc_buf = iov[num_iovecs-1].iov_base;
  }
 
  rc = SendReceive2(xid, ses, iov, num_iovecs, &resp_buftype, 0);
@@ -1211,6 +1214,8 @@ SMB2_open(const unsigned int xid, struct cifs_open_parms *oparms, __le16 *path,
   *oplock = rsp->OplockLevel;
 creat_exit:
  kfree(copy_path);
+ kfree(lc_buf);
+ kfree(dhc_buf);
  free_rsp_buf(resp_buftype, rsp);
  return rc;
 }
-- 
1.7.1